activation/atxdb.go

```diff
@@ -435,27 +438,34 @@ func (db *DB) SyntacticallyValidateAtx(atx *types.ActivationTx) error {
 // ContextuallyValidateAtx ensures that the previous ATX referenced is the last known ATX for the referenced miner ID.
 // If a previous ATX is not referenced, it validates that indeed there's no previous known ATX for that miner ID.
 func (db *DB) ContextuallyValidateAtx(atx *types.ActivationTxHeader) error {
-	if atx.PrevATXID != *types.EmptyATXID {
-		lastAtx, err := db.GetNodeLastAtxID(atx.NodeID)
-		if err != nil {
-			db.log.With().Error("could not fetch node last ATX", atx.ID(),
-				log.FieldNamed("atx_node_id", atx.NodeID), log.Err(err))
-			return fmt.Errorf("could not fetch node last ATX: %v", err)
-		}
-		// last atx is not the one referenced
-		if lastAtx != atx.PrevATXID {
-			return fmt.Errorf("last atx is not the one referenced")
-		}
-	} else {
-		lastAtx, err := db.GetNodeLastAtxID(atx.NodeID)
-		if _, ok := err.(ErrAtxNotFound); err != nil && !ok {
-			db.log.Error("fetching ATX ids failed: %v", err)
-			return err
-		}
-		if err == nil { // we found an ATX for this node ID, although it reported no prevATX -- this is invalid
-			return fmt.Errorf("no prevATX reported, but other ATX with same nodeID (%v) found: %v",
-				atx.NodeID.ShortString(), lastAtx.ShortString())
+	if atx.PrevATXID == *types.EmptyATXID {
+		return fmt.Errorf("no prevATX reported")
+	}
+
+	lastAtx, err := db.GetNodeLastAtxID(atx.NodeID)
+	if _, ok := err.(ErrAtxNotFound); ok {
+		if atx.PrevATXID != db.goldenATXID {
+			// we found no ATXs for this node ID, although it reported no golden ATX as prevATX -- this is invalid
+			return fmt.Errorf("other ATXs not found, but golden ATX not reported as prevATX")
 		}
+
+		// we found no ATXs for this node ID, and it reported golden ATX as prevATX -- this is valid
+		return nil
+	} else if err != nil {
+		db.log.With().Error("could not fetch node last ATX", atx.ID(),
+			log.FieldNamed("atx_node_id", atx.NodeID), log.Err(err))
+		return fmt.Errorf("could not fetch node last ATX: %v", err)
+	}
+
+	if atx.PrevATXID == db.goldenATXID {
```

Comment:
this check is redundant, since the next check covers this case, if we've found an ATX in our DB this is the ATX that this node should point to

Reply:
Yes, it's redundant if we don't need a separate error for this case. Removed.